flusher.go

```diff
@@ -82,85 +143,192 @@ func (s *Server) Flush(interval time.Duration, metricLimit int) {
 	wg.Wait()
 	s.statsd.TimeInMilliseconds("flush.total_duration_ns", float64(time.Now().Sub(flushStart).Nanoseconds()), nil, 1.0)
 
-	s.statsd.Count("flush.error_total", 0, nil, 1.0) // make sure this metric is not sparse
 	s.logger.WithField("metrics", len(finalMetrics)).Info("Completed flush to Datadog")
 }
 
 func (s *Server) flushPart(metricSlice []DDMetric, wg *sync.WaitGroup) {
 	defer wg.Done()
-
-	cstart := time.Now()
-	var reqBody bytes.Buffer
-	compressor := zlib.NewWriter(&reqBody)
-	encoder := json.NewEncoder(compressor)
-	err := encoder.Encode(map[string][]DDMetric{
+	s.postHelper(fmt.Sprintf("%s/api/v1/series?api_key=%s", s.DDHostname, s.DDAPIKey), map[string][]DDMetric{
 		"series": metricSlice,
-	})
-	if err != nil {
-		s.statsd.Count("flush.error_total", 1, []string{"cause:json"}, 1.0)
-		s.logger.WithError(err).Error("Error rendering JSON request body")
+	}, "flush")
+}
+
+func (s *Server) flushForward(wms []WorkerMetrics) {
+	jmLength := 0
+	for _, wm := range wms {
+		jmLength += len(wm.histograms)
+		jmLength += len(wm.sets)
+		jmLength += len(wm.timers)
+	}
+
+	jsonMetrics := make([]JSONMetric, 0, jmLength)
+	for _, wm := range wms {
+		for _, histo := range wm.histograms {
+			jm, err := histo.Export()
+			if err != nil {
+				s.logger.WithFields(logrus.Fields{
+					logrus.ErrorKey: err,
+					"type":          "histogram",
+					"name":          histo.name,
+				}).Error("Could not export metric")
+				continue
+			}
+			jsonMetrics = append(jsonMetrics, jm)
+		}
+		for _, set := range wm.sets {
+			jm, err := set.Export()
+			if err != nil {
+				s.logger.WithFields(logrus.Fields{
+					logrus.ErrorKey: err,
+					"type":          "set",
+					"name":          set.name,
+				}).Error("Could not export metric")
+				continue
+			}
+			jsonMetrics = append(jsonMetrics, jm)
+		}
+		for _, timer := range wm.timers {
+			jm, err := timer.Export()
+			if err != nil {
+				s.logger.WithFields(logrus.Fields{
+					logrus.ErrorKey: err,
+					"type":          "timer",
+					"name":          timer.name,
+				}).Error("Could not export metric")
+				continue
+			}
+			jsonMetrics = append(jsonMetrics, jm)
+		}
+	}
+
+	s.statsd.Gauge("forward.post_metrics_total", float64(len(jsonMetrics)), nil, 1.0)
```

Comment:
I prefer to use the length directly and treat the precomputation as an optimization only